airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/async_retriever.py

```diff
@@ -0,0 +1,113 @@
+    stream_slicer: StreamSlicer = field(default_factory=lambda: SinglePartitionRouter(parameters={}))
+
+    def __post_init__(self, parameters: Mapping[str, Any]) -> None:
+        self._job_orchestrator_factory = self.job_orchestrator_factory
+        self.__job_orchestrator: Optional[AsyncJobOrchestrator] = None
+        self._parameters = parameters
+
+    @property
+    def state(self) -> StreamState:
+        """
+        As a first iteration for sendgrid, there is no state to be managed
+        """
+        return {}
+
+    @state.setter
+    def state(self, value: StreamState) -> None:
+        """
+        As a first iteration for sendgrid, there is no state to be managed
+        """
+        pass
+
+    @property
+    def _job_orchestrator(self) -> AsyncJobOrchestrator:
+        if not self.__job_orchestrator:
+            raise AirbyteTracedException(
+                message="Invalid state within AsyncJobRetriever. Please contact Airbyte Support",
+                internal_message="AsyncPartitionRepository is expected to be accessed only after `stream_slices`",
+                failure_type=FailureType.system_error,
+            )
+
+        return self.__job_orchestrator
+
+    def _get_stream_state(self) -> StreamState:
+        """
+        Gets the current state of the stream.
+
+        Returns:
+            StreamState: Mapping[str, Any]
+        """
+
+        return self.state
+
+    def _validate_and_get_stream_slice_partition(self, stream_slice: Optional[StreamSlice] = None) -> AsyncPartition:
+        """
+        Validates the stream_slice argument and returns the partition from it.
+
+        Args:
+            stream_slice (Optional[StreamSlice]): The stream slice to validate and extract the partition from.
+
+        Returns:
+            AsyncPartition: The partition extracted from the stream_slice.
+
+        Raises:
+            AirbyteTracedException: If the stream_slice is not an instance of StreamSlice or if the partition is not present in the stream_slice.
+
+        """
+        if not isinstance(stream_slice, StreamSlice) or "partition" not in stream_slice.partition:
+            raise AirbyteTracedException(
+                message="Invalid arguments to AsyncJobRetriever.read_records: stream_slice is no optional. Please contact Airbyte Support",
+                failure_type=FailureType.system_error,
+            )
+        return stream_slice["partition"]  # type: ignore  # stream_slice["partition"] has been added as an AsyncPartition as part of stream_slices
+
+    def stream_slices(self) -> Iterable[Optional[StreamSlice]]:
+        slices = self.stream_slicer.stream_slices()
+        self.__job_orchestrator = self._job_orchestrator_factory(slices)
+
+        for completed_partition in self._job_orchestrator.create_and_get_completed_partitions():
+            yield StreamSlice(
+                partition=dict(completed_partition.stream_slice.partition) | {"partition": completed_partition},
```

Comment:
@brianjlai I thought this might be interesting for you. This doesn't work very well for concurrent because:
* The slice is not json serializable as it is of object `StreamSlice` and it'll fail [here](https://github.com/airbytehq/airbyte/blob/f23ae56693c931e36028a61ea1b11ffdd25762e2/airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/adapters.py#L276)
* Even if we were to cast it as a dict, this would fail because `completed_partition` is of type `AsyncJob` which is also not serializable

Reply:
is there a real blocker to making `AsyncJob` serializable?

maybe hot take: partitions should be serializable else there's no way to pass the context from a thread to another

Reply:
My understanding of how the `json` library works is that we would need to provide a `default`

[doc](https://docs.python.org/3/library/json.html#json.dump) says: `If specified, default should be a function that gets called for objects that can’t otherwise be serialized. It should return a JSON encodable version of the object or raise a [TypeError](https://docs.python.org/3/library/exceptions.html#TypeError). If not specified, [TypeError](https://docs.python.org/3/library/exceptions.html#TypeError) is raised.`

In our case, I don't know if we could:
* assume that if we can't serialize it with the standard encoder, we use the string version
* ask users that puts an object in the slice to implement a `__stream_slice_serialize__` method

I think I would prefer the second option but haven't put too much thoughts into it yet